server/serverAPI/controller/AuthenticationController.ts

```diff
@@ -101,6 +118,72 @@ export default class AuthenticationController extends BaseUserController {
         }
     }
 
+    confirmVerificationCode = async (req: Request, res: Response) => {
+        let username = req.body?.username;
+        let actualCode = req.body?.verificationCode;
+
+        let requestedCode = AuthenticationController.verificationCodesMap.get(username);
+
+        if (requestedCode === undefined) {
+            return this.sendError(400, res, "Verification code is either expired or not issued.");
+        }
+
+        if (actualCode !== requestedCode) {
+            return this.sendError(400, res, "Verification code is invalid.");
+        }
+
+        try {
+            let user = await this.requestGet(new Map([["username", username]]), res);
+
+            user.isVerified = true;
+
+            await this.requestUpdate(user.username, user, res);
+
+            return this.sendSuccess(200, res, "Account has been verified.");
+        } catch (e) {
+            return e;
+        }
+    }
+
+    sendVerificationCode = async (req: Request, res: Response) => {
+        let email = "";
+        let username = req.body?.username;
+
+        try {
+            let user = await this.requestGet(new Map([["username", username]]), res);
+            email = user.email;
+        } catch(e) {
+            return e;
+        }
+
+        let verificationCode = Random.getRandomNumber(this.minVerificationCode, this.maxVerificationCode);
+
+        let emailVerificationSchema = new EmailVerificationTemplateSchema(
+            username,
+            verificationCode
+        );
+
+        try {
+            await this.verifySchema(emailVerificationSchema, res);
+            await this.requestGet(new Map([["username", username]]), res)
+        } catch (e) {
+            return e;
+        }
+
+        AuthenticationController.verificationCodesMap.set(username, verificationCode);
```

Comment:
Should there be some limit on number of verification codes that can be sent in a given time interval?

Reply:
honestly, have no idea how to do it nicely...

Reply:
In addition to storing the verification code, you can also store the time that the last code was generated. If the difference in the time is below some threshold, you cannot create a new one. Or alternatively, you can store the number of requests generated, increment it when you create a new code, and decrement it after some time threshold using a callback. Whichever you prefer